engines/wintermute/ad/ad_entity.cpp

```diff
@@ -838,6 +840,78 @@ bool AdEntity::scCallMethod(ScScript *script, ScStack *stack, ScStack *thisStack
 		return STATUS_OK;
 	}
 
+#if defined(ENABLE_KINJAL) || defined(ENABLE_HEROCRAFT)
+	//////////////////////////////////////////////////////////////////////////
+	// [WME Kinjal 1.4] SetBeforeEntity / SetAfterEntity
+	// Usage at HeroCraft games: ent.SetBeforeEntity("redDuskaEntity")
+	//////////////////////////////////////////////////////////////////////////
+	else if (strcmp(name, "SetBeforeEntity") == 0 || strcmp(name, "SetAfterEntity") == 0) {
+		stack->correctParams(1);
+		const char *layerName = stack->pop()->getString();
+
+		if (strcmp(getName(), layerName) == 0) {
+			warning("%s(%s): source and target has the same name", name, layerName);
+			stack->pushBool(false);
+			return STATUS_OK;
+		}
+
+		for (uint32 i = 0; i < ((AdGame *)_gameRef)->_scene->_layers.size(); i++) {
+			AdLayer *layer = ((AdGame *)_gameRef)->_scene->_layers[i];
+			for (uint32 j = 0; j < layer->_nodes.size(); j++) {
+				if (layer->_nodes[j]->_type == OBJECT_ENTITY && this == layer->_nodes[j]->_entity) {
+					for (uint32 k = 0; k < layer->_nodes.size(); k++) {
+						if (layer->_nodes[k]->_type == OBJECT_ENTITY && strcmp(layer->_nodes[k]->_entity->getName(), layerName) == 0) {
+							if (j < k && strcmp(name, "SetBeforeEntity") == 0) {
+								k--;
+							} else if (j > k && strcmp(name, "SetAfterEntity") == 0) {
+								k++;
+							}
```

Comment:
Comparing strings in each for loop iteration is costly. I suggest you store the result of the comparison in a boolean variable outside the for loops, and use that instead, i.e. something like:

    bool isSetBeforeEntity = !strcmp(name, "SetBeforeEntity");    // before the for loops
    bool isSetAfterEntity = !strcmp(name, "SetAfterEntity");    // before the for loops
    ....
    if (j < k && isSetBeforeEntity)
        k--;
    else if (j > k && isSetAfterEntity)
        k++;

Reply:
There is `return` statement after those `strcmp`s, so they are actually called only once.
Added some comments since this function is too complex and confusing.